### ISSUE TITLE
Tensor.feature_dim and Tensor.feature_dim_axis

### ISSUE THREAD

**Issue by albertz** [rwth-i6/returnn#1273]
The question is whether to use them in new code, esp whether it is used in the new frontend API (https://github.com/rwth-i6/returnn/issues/1264, https://github.com/rwth-i6/returnn/issues/1120), or whether they should be avoided. See also here for some initial comments: https://github.com/rwth-i6/returnn_common/issues/252#issuecomment-1460933674

Some arguments to use them:

- The code might be too verbose otherwise. It potentially can simplify a lot of code.
- There are already a lot of usages via returnn-common `nn`. When the frontend API becomes the underlying core behind returnn-common `nn`, it would break a lot of code.

Some arguments against them:

- There might be ambiguous cases, where it is not clear what `feature_dim` is set to. We should try to make sure such cases do not exist, and `feature_dim` is just None after any ambiguous op, like `split_dims`, or when the dim was removed, like `reduce`. But for example, what about `dot`? Would `dot` handle it, or is it up to `Linear`? Similarly `conv`, `lstm`, etc.
- There might be code which works if `feature_dim` is set to one dim and then breaks when it is set otherwise, and this is maybe difficult to debug. It might make it harder to reason about some code. It adds another thing to think about.
- It can make some internal code more complicated. E.g. inside of `dot`, we need some extra handling to correctly set the returned `feature_dim`.

**Comment by albertz** [comment]
Currently, `feature_dim` is internally used (as default axis, or just always) in:

* `cross_entropy` and other loss functions
* `concat_features`
* `gating`
